File: kotlin/bc-components/src/main/kotlin/com/blockchaincommons/bccomponents/Digest.kt
package com.blockchaincommons.bccomponents

import com.blockchaincommons.bccrypto.sha256
import com.blockchaincommons.bctags.TAG_DIGEST
import com.blockchaincommons.bcur.UR
import com.blockchaincommons.bcur.URCodable
import com.blockchaincommons.bcur.URDecodable
import com.blockchaincommons.bcur.UREncodable
import com.blockchaincommons.dcbor.Cbor
import com.blockchaincommons.dcbor.CborTaggedCodable
import com.blockchaincommons.dcbor.CborTaggedDecodable
import com.blockchaincommons.dcbor.CborTaggedEncodable
import com.blockchaincommons.dcbor.CborTaggedUtils
import com.blockchaincommons.dcbor.Tag
import com.blockchaincommons.dcbor.tagsForValues

/**
 * A cryptographically secure digest, implemented with SHA-256.
 *
 * A [Digest] represents the cryptographic hash of some data. SHA-256 is
 * used, producing a 32-byte hash value. Digests are used throughout the
 * library for data verification and as unique identifiers derived from data.
 */
class Digest private constructor(private val data: ByteArray) :
    DigestProvider,
    Comparable<Digest>,
    CborTaggedCodable,
    URCodable {

    init {
        require(data.size == DIGEST_SIZE) {
            "Digest data must be exactly $DIGEST_SIZE bytes, got ${data.size}"
        }
    }

    /** Returns a copy of the underlying 32-byte digest data. */
    fun data(): ByteArray = data.copyOf()

    /** Returns the digest bytes as a read-only slice. */
    fun asBytes(): ByteArray = data.copyOf()

    /** Validates that this digest matches the SHA-256 hash of [image]. */
    fun validate(image: ByteArray): Boolean = this == fromImage(image)

    /** The digest as a 64-character lowercase hexadecimal string. */
    val hex: String get() = data.toHexString()

    /** Returns the first four bytes of the digest as a hexadecimal string. */
    fun shortDescription(): String = data.copyOfRange(0, 4).toHexString()

    /** Returns the digest bytes as a new [ByteArray]. */
    fun toByteArray(): ByteArray = data.copyOf()

    // -- DigestProvider --

    override fun digest(): Digest = this

    // -- Comparable --

    override fun compareTo(other: Digest): Int {
        for (i in data.indices) {
            val cmp = (data[i].toInt() and 0xFF) - (other.data[i].toInt() and 0xFF)
            if (cmp != 0) return cmp
        }
        return 0
    }

    // -- equals / hashCode --

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is Digest) return false
        return data.contentEquals(other.data)
    }

    override fun hashCode(): Int = data.contentHashCode()

    // -- toString / debug --

    override fun toString(): String = "Digest($hex)"

    // -- CBOR --

    override fun cborTags(): List<Tag> = tagsForValues(listOf(TAG_DIGEST))

    override fun untaggedCbor(): Cbor = Cbor.fromByteString(data)

    // -- Companion --

    companion object {
        const val DIGEST_SIZE: Int = 32

        /** Creates a digest from exactly [DIGEST_SIZE] bytes. */
        fun fromData(data: ByteArray): Digest {
            require(data.size == DIGEST_SIZE) {
                "Digest data must be exactly $DIGEST_SIZE bytes, got ${data.size}"
            }
            return Digest(data.copyOf())
        }

        /**
         * Creates a digest from a byte array, returning an error if the
         * length is not [DIGEST_SIZE].
         */
        fun fromDataChecked(data: ByteArray): Digest {
            if (data.size != DIGEST_SIZE) {
                throw BcComponentsException.invalidSize("digest", DIGEST_SIZE, data.size)
            }
            return Digest(data.copyOf())
        }

        /** Creates a digest by computing the SHA-256 hash of [image]. */
        fun fromImage(image: ByteArray): Digest = Digest(sha256(image))

        /**
         * Creates a digest by concatenating the given byte parts and
         * computing their SHA-256 hash.
         */
        fun fromImageParts(parts: List<ByteArray>): Digest {
            val totalSize = parts.sumOf { it.size }
            val buf = ByteArray(totalSize)
            var offset = 0
            for (part in parts) {
                part.copyInto(buf, offset)
                offset += part.size
            }
            return fromImage(buf)
        }

        /**
         * Creates a digest by concatenating the data of the given digests
         * and computing their SHA-256 hash.
         */
        fun fromDigests(digests: List<Digest>): Digest {
            val buf = ByteArray(digests.size * DIGEST_SIZE)
            var offset = 0
            for (d in digests) {
                d.data.copyInto(buf, offset)
                offset += DIGEST_SIZE
            }
            return fromImage(buf)
        }

        /**
         * Creates a digest from a hexadecimal string.
         *
         * @throws IllegalArgumentException if the string is not exactly 64
         *   hex digits.
         */
        fun fromHex(hex: String): Digest = fromData(hex.hexToByteArray())

        /**
         * Validates [image] against an optional digest.
         *
         * Returns `true` if [digest] is `null` or if it matches the image's
         * SHA-256 hash. Returns `false` otherwise.
         */
        fun validate(image: ByteArray, digest: Digest?): Boolean =
            digest?.validate(image) ?: true

        /** Decodes a [Digest] from untagged CBOR (a byte string). */
        fun fromUntaggedCbor(cbor: Cbor): Digest {
            val bytes = cbor.tryByteStringData()
            return fromDataChecked(bytes)
        }

        /** Decodes a [Digest] from tagged CBOR. */
        fun fromTaggedCbor(cbor: Cbor): Digest =
            CborTaggedUtils.fromTaggedCbor(
                cbor,
                tagsForValues(listOf(TAG_DIGEST)),
            ) { fromUntaggedCbor(it) }

        /** Decodes a [Digest] from tagged CBOR binary data. */
        fun fromTaggedCborData(data: ByteArray): Digest =
            CborTaggedUtils.fromTaggedCborData(
                data,
                tagsForValues(listOf(TAG_DIGEST)),
            ) { fromUntaggedCbor(it) }

        /** Decodes a [Digest] from a UR. */
        fun fromUr(ur: UR): Digest {
            ur.checkType("digest")
            return fromUntaggedCbor(ur.cbor)
        }

        /** Decodes a [Digest] from a UR string. */
        fun fromUrString(urString: String): Digest =
            fromUr(UR.fromUrString(urString))
    }
}

File: csharp/BCComponents/BCComponents/Digest.cs
using BlockchainCommons.BCCrypto;
using BlockchainCommons.BCTags;
using BlockchainCommons.BCUR;
using BlockchainCommons.DCbor;

namespace BlockchainCommons.BCComponents;

/// <summary>
/// A cryptographically secure digest, implemented with SHA-256.
/// </summary>
/// <remarks>
/// A <see cref="Digest"/> represents the cryptographic hash of some data.
/// SHA-256 is used, producing a 32-byte hash value. Digests are used
/// throughout the library for data verification and as unique identifiers
/// derived from data.
/// </remarks>
public sealed class Digest : IEquatable<Digest>, IComparable<Digest>, ICborTagged, ICborTaggedEncodable, ICborTaggedDecodable, IUREncodable, IURDecodable, IDigestProvider
{
    /// <summary>The size of a digest in bytes.</summary>
    public const int Size = 32;

    private readonly byte[] _data;

    private Digest(byte[] data)
    {
        _data = data;
    }

    /// <summary>
    /// Creates a new digest from a 32-byte array.
    /// </summary>
    /// <param name="data">Exactly 32 bytes of digest data.</param>
    /// <returns>A new <see cref="Digest"/>.</returns>
    /// <exception cref="ArgumentException">Thrown if <paramref name="data"/> is not exactly 32 bytes.</exception>
    public static Digest FromData(byte[] data)
    {
        if (data.Length != Size)
            throw BCComponentsException.InvalidSize("digest", Size, data.Length);
        var copy = new byte[Size];
        Array.Copy(data, copy, Size);
        return new Digest(copy);
    }


    /// <summary>
    /// Creates a new digest by hashing the given image with SHA-256.
    /// </summary>
    /// <param name="image">The data to hash.</param>
    /// <returns>A new <see cref="Digest"/> containing the SHA-256 hash.</returns>
    public static Digest FromImage(byte[] image)
    {
        return new Digest(Hash.Sha256(image));
    }

    /// <summary>
    /// Creates a new digest by concatenating image parts and hashing with SHA-256.
    /// </summary>
    /// <param name="imageParts">The data parts to concatenate and hash.</param>
    /// <returns>A new <see cref="Digest"/>.</returns>
    public static Digest FromImageParts(byte[][] imageParts)
    {
        int totalLen = 0;
        foreach (var part in imageParts)
            totalLen += part.Length;

        var buf = new byte[totalLen];
        int offset = 0;
        foreach (var part in imageParts)
        {
            Array.Copy(part, 0, buf, offset, part.Length);
            offset += part.Length;
        }
        return FromImage(buf);
    }

    /// <summary>
    /// Creates a new digest from an array of digests by concatenating their data and hashing.
    /// </summary>
    /// <param name="digests">The digests to combine.</param>
    /// <returns>A new <see cref="Digest"/> computed from the concatenated digest data.</returns>
    public static Digest FromDigests(Digest[] digests)
    {
        var buf = new byte[digests.Length * Size];
        for (int i = 0; i < digests.Length; i++)
        {
            Array.Copy(digests[i]._data, 0, buf, i * Size, Size);
        }
        return FromImage(buf);
    }

    /// <summary>Gets a copy of the digest data as a 32-byte array.</summary>
    public byte[] Data => (byte[])_data.Clone();

    /// <summary>Returns the digest data as a read-only span.</summary>
    /// <returns>A read-only span over the digest bytes.</returns>
    public ReadOnlySpan<byte> AsBytes() => _data;

    /// <summary>
    /// Validates that this digest matches the SHA-256 hash of the given image.
    /// </summary>
    /// <param name="image">The data to validate against.</param>
    /// <returns><c>true</c> if the digest matches the hash of <paramref name="image"/>.</returns>
    public bool Validate(byte[] image)
    {
        return Equals(FromImage(image));
    }

    /// <summary>
    /// Creates a new digest from a hexadecimal string.
    /// </summary>
    /// <param name="hex">A 64-character hexadecimal string.</param>
    /// <returns>A new <see cref="Digest"/>.</returns>
    /// <exception cref="FormatException">Thrown if the hex string is invalid.</exception>
    /// <exception cref="BCComponentsException">Thrown if the decoded data is not exactly 32 bytes.</exception>
    public static Digest FromHex(string hex)
    {
        var data = Convert.FromHexString(hex);
        return FromData(data);
    }

    /// <summary>Gets the digest data as a lowercase hexadecimal string.</summary>
    public string Hex => Convert.ToHexString(_data).ToLowerInvariant();

    /// <summary>Returns a short description of the digest (first 4 bytes as hex).</summary>
    /// <returns>An 8-character hexadecimal string.</returns>
    public string ShortDescription() => Convert.ToHexString(_data, 0, 4).ToLowerInvariant();

    /// <summary>
    /// Validates the given data against an optional digest.
    /// </summary>
    /// <param name="image">The data to validate.</param>
    /// <param name="digest">The digest to validate against, or <c>null</c>.</param>
    /// <returns>
    /// <c>true</c> if <paramref name="digest"/> is <c>null</c> or if it matches
    /// the SHA-256 hash of <paramref name="image"/>.
    /// </returns>
    public static bool ValidateOpt(byte[] image, Digest? digest)
    {
        return digest == null || digest.Validate(image);
    }

    // --- IDigestProvider ---

    /// <summary>Returns itself, as a Digest is already a digest.</summary>
    /// <returns>This digest.</returns>
    Digest IDigestProvider.GetDigest() => this;

    // --- ICborTagged ---

    /// <summary>Returns the CBOR tags associated with the Digest type.</summary>
    public static IReadOnlyList<Tag> CborTags => GlobalTags.TagsForValues(BcTags.TagDigest);

    // --- ICborTaggedEncodable ---

    /// <summary>Returns the untagged CBOR representation (a byte string).</summary>
    public Cbor UntaggedCbor() => Cbor.ToByteString(_data);

    /// <summary>Returns the tagged CBOR representation.</summary>
    public Cbor TaggedCbor() => Cbor.ToTaggedValue(CborTags[0], UntaggedCbor());

    /// <summary>Returns the tagged CBOR representation (used by IUREncodable).</summary>
    public Cbor ToCbor() => TaggedCbor();

    // --- ICborTaggedDecodable ---

    /// <summary>
    /// Decodes a <see cref="Digest"/> from an untagged CBOR byte string.
    /// </summary>
    /// <param name="cbor">The untagged CBOR value.</param>
    /// <returns>A new <see cref="Digest"/>.</returns>
    /// <exception cref="CborException">Thrown if the CBOR is not a valid byte string.</exception>
    /// <exception cref="BCComponentsException">Thrown if the data is not exactly 32 bytes.</exception>
    public static Digest FromUntaggedCbor(Cbor cbor)
    {
        var data = cbor.TryIntoByteString();
        return FromData(data);
    }

    /// <summary>
    /// Decodes a <see cref="Digest"/> from a tagged CBOR value.
    /// </summary>
    /// <param name="cbor">The tagged CBOR value.</param>
    /// <returns>A new <see cref="Digest"/>.</returns>
    /// <exception cref="CborException">Thrown if the CBOR tag does not match or the data is invalid.</exception>
    public static Digest FromTaggedCbor(Cbor cbor)
    {
        foreach (var tag in CborTags)
        {
            try
            {
                var item = cbor.TryIntoExpectedTaggedValue(tag);
                return FromUntaggedCbor(item);
            }
            catch (CborWrongTagException) { }
            catch (CborWrongTypeException) { }
        }
        throw new CborWrongTypeException();
    }

    // --- IEquatable<Digest> ---

    /// <inheritdoc/>
    public bool Equals(Digest? other)
    {
        if (other is null) return false;
        return _data.AsSpan().SequenceEqual(other._data);
    }

    /// <inheritdoc/>
    public override bool Equals(object? obj) => obj is Digest d && Equals(d);

    /// <inheritdoc/>
    public override int GetHashCode()
    {
        var hash = new HashCode();
        foreach (var b in _data)
            hash.Add(b);
        return hash.ToHashCode();
    }

    /// <summary>Tests equality of two Digest instances.</summary>
    public static bool operator ==(Digest? left, Digest? right) =>
        left is null ? right is null : left.Equals(right);

    /// <summary>Tests inequality of two Digest instances.</summary>
    public static bool operator !=(Digest? left, Digest? right) => !(left == right);

    // --- IComparable<Digest> ---

    /// <summary>
    /// Compares this digest to another lexicographically by their underlying bytes.
    /// </summary>
    /// <param name="other">The other digest to compare to.</param>
    /// <returns>A negative, zero, or positive value indicating relative order.</returns>
    public int CompareTo(Digest? other)
    {
        if (other is null) return 1;
        return _data.AsSpan().SequenceCompareTo(other._data);
    }

    // --- Display ---

    /// <inheritdoc/>
    public override string ToString() => $"Digest({Hex})";

    /// <summary>Returns a byte array copy of this digest's data.</summary>
    /// <returns>A 32-byte array.</returns>
    public byte[] ToByteArray() => Data;
}
